strawberryfields/api/result.py

```diff
@@ -107,6 +103,72 @@ def state(self):
             raise AttributeError("The state is undefined for a stateless computation.")
         return self._state
 
+    @staticmethod
+    def combine_samples(samples_list, mode_order):
+        """Combine samples measured at different times into one nested array.
+
+        Args:
+            samples_list (list[array, tensor]): the sample measurements in a list
+            mode_order (list[array]): The nested mode order for the measured modes. Must have the
+                same structure as samples_list.
+
+        Returns:
+            array: the samples in ascending mode order with shape ``(shots, measured_modes)``
+        """
+
+        shots = len(samples_list[0])
+
+        mode_order_flat = np.hstack(mode_order)
+        len_modes = len(np.unique(mode_order_flat))
+
+        # remove duplicate mode-measurements and only return the modes from the last measurement
+        duplicate_mode_measurements = len(mode_order_flat) != len_modes
+        if duplicate_mode_measurements:
+            # change to list to allow updates via reference, and split samples into shots
+            mode_order_list = [list(m) for m in mode_order]
+            samples_shot_list = [[list(s[i]) for s in samples_list] for i in range(shots)]
+
+            combos = it.combinations(range(len(mode_order_list)), 2)
+
+            # go through each combination of mode orders (one per measurement) and delete
+            # the first occurence of any duplicates; also delete corresponding sample(s)
+            for a, b in it.combinations(mode_order_list, 2):
+                intersect, *idx = np.intersect1d(a, b, return_indices=True)
+                k = combos.__next__()[0]
+                for i, inter in enumerate(intersect):
+                    a.remove(inter)
+                    for j in range(shots):
+                        # remove last sample in shot array so it doesn't disturb the indices
+                        s_val = samples_shot_list[j][k][idx[k][len(intersect) - i - 1]]
+                        samples_shot_list[j][k].remove(s_val)
+
+            # reshape into a list of arrays, ignoring empty ones
+            mode_order = [np.array(m) for m in mode_order_list if m]
+            samples_list = [np.array([s]) for s in zip(*samples_shot_list) if s[0]]
+
+            shapes = np.cumsum([len(m) for m in mode_order])
```

Comment:
This could perhaps be formulated as follows to break down the intermediate steps a little bit (had a quick chat with Theo on this, curious if any solution would be preferred overall):
```python
# 1. gather the indices of where removal needs to happen
modes_to_remove = []
modes_already = []
for idx, modes in enumerate(reversed(mode_order)):
    removal_indices = []
    for mode in modes:            
        if mode in modes_already:
            removal_indices.append(int(np.argwhere(modes==mode)))
        else:
            modes_already.append(mode)
    modes_to_remove.append((idx, np.array(removal_indices)))

# 2. Remove
for reversed_idx, removal_indices in modes_to_remove:
    idx = len(mode_order) - reversed_idx -1
    if removal_indices.size != 0:
        samples_list[idx] = np.delete(samples_list[idx], removal_indices, axis=1)
        mode_order[idx] = np.delete(mode_order[idx], removal_indices, axis=0)
print(samples_list, mode_order)
```

Reply:
Thanks @antalszava. I think this is a nice solution as well, although I'm not sure which I prefer. I think it's better if someone could look at this with fresh eyes. @josh146 @co9olguy

Reply:
@thisac I wonder if it might make more sense to address this at the _engine_ level rather than here. This would

* avoid this quite complicated logic, and allow it to be reverted to how it was before

* seems cleaner.

For example, in the `eng.run()` method, we have

```python
values = []
for cmd in prog.circuit:
    try:
        # try to apply it to the backend
        # try to apply it to the backend and, if op is a measurement, store it in values
        # NOTE we could also handle storing measured vals here
        val = cmd.op.apply(cmd.reg, self.backend, **kwargs)
        
        if val is not None:
            values.append(val)
```

Perhaps we could make `values` a _dictionary_, indexed by mode number via `cmd.reg`? That way, when a mode is measured again, the values returned overwrite the previous values automatically.

Reply:
@josh146 That's actually a nice idea. I feel like that maybe should've been the way to go from the start. :smile: 

I actually started to think about whether it's actually worth having all of this logic in a separate `Results.combine_samples` function. Another solution, which looks a bit more like the earlier one, would be having all of the (much simpler) logic inside `engine.run`. In that case both `Results.combine_samples`, etc. could be removed, since e.g. the code below would return the samples correctly right away.

This seems to pass all the tests; where I've only changed the way the values or stored, as well as added a much simpler `_combining_samples` method:

```python
def _run_program(self, prog, **kwargs):
    applied = []
    values = {}
    batches = self.backend_options.get("batch_size", 0)
    for cmd in prog.circuit:
        try:
            # try to apply it to the backend and, if op is a measurement, store it in values
            val = cmd.op.apply(cmd.reg, self.backend, **kwargs)
            if val is not None:
                for i, r in enumerate(cmd.reg):
                    if batches:
                        values[r.ind] = np.array(val[:, :, i])
                    else:
                        values[r.ind] = val[:, i]

            applied.append(cmd)

        except NotApplicableError:
            # command is not applicable to the current backend type
            raise NotApplicableError(
                "The operation {} cannot be used with {}.".format(cmd.op, self.backend)
            ) from None

        except NotImplementedError:
            # command not directly supported by backend API
            raise NotImplementedError(
                "The operation {} has not been implemented in {} for the arguments {}.".format(
                    cmd.op, self.backend, kwargs
                )
            ) from None

    samples = self._combine_samples(values, kwargs["shots"], batches)

    return applied, samples
```
The `_combine_samples` method above (which can probably be written in a nicer way as well) would be:

```python
def _combine_samples(self, values, shots, batches):
    if values == {}:
        return []

    samples = np.array([[i[j] for _, i in sorted(values.items())] for j in range(shots)])

    # pylint: disable=import-outside-toplevel
    if self.backend_name == "tf":
        from tensorflow import convert_to_tensor

        if batches:
            samples = convert_to_tensor(
                np.array(
                    [
                        [[i[b][j] for _, i in sorted(values.items())] for j in range(shots)]
                        for b in range(batches)
                    ]
                )
            )
        else:
            samples = convert_to_tensor(samples)

    return samples
```

This way, there is not that much extra logic trying to filter out and combine the samples afterwards, and I feel like it's a bit clearer to understand what's going on. But then we might as well have just kept the old way of getting the samples from the register instead. :thinking: